### PR TITLE
Update the Aravis Camera module to have better robustness and work wi…

### PR DESCRIPTION
**Pull request by TrentHouliston**
This PR updates the Camera module to the latest version of aravis
It also adds newer robustness and PTP code (in case you ever start using GigE cameras)

### Pre-PR Checklist:

Have you
- [x] Updated NUbook if necessary (add link to NUbook PR here)
- [x] Added/updated tests for your changes, including regression tests for bug fixes
- [x] Updated relevant module READMEs
- [x] Added/modified [documentation directives](https://nubook.nubots.net/guides/general/code-conventions#documentation) in relevant code
- [x] Added a descriptive title and relevant labels to the PR

### CODE REVIEW

File: module/input/Camera/src/Camera.cpp [hunk 1]
```diff
@@ -462,12 +228,408 @@ namespace module::input {
...
+            throw std::runtime_error(fmt::format("{} camera with serial number {} not found", name, serial_number));
+        };
+
+        // Find the device, and if we can't check for new devices
+        int device_no = 0;
+        try {
+            device_no = find_camera(name, serial_number);
+        }
+        catch (const std::runtime_error&) {
+            arv_update_device_list();
```
Comment (CMurtagh-LGTM) [inline module/input/Camera/src/Camera.cpp:273]: Is this rare enough that the try-catch is more performant?

File: module/input/Camera/src/Camera.cpp [hunk 2]
```diff
@@ -462,12 +228,408 @@ namespace module::input {
...
+        }
+
+        reactor.log<NUClear::INFO>(fmt::format("Connected to the {} camera ({})", name, device_description));
+
+        // Return camera context
+        return CameraContext{reactor,
+                             config,
+                             name,
+                             0,  // fourcc is set later
+                             XXH32(name.data(), name.size(), 0x4E55436C),
```
Comment (CMurtagh-LGTM) [inline module/input/Camera/src/Camera.cpp:321]: Magic number.

File: module/input/Camera/src/Camera.cpp [hunk 3]
```diff
@@ -462,12 +228,408 @@ namespace module::input {
...
+
+        // Apply the region to the camera
+        arv::camera_set_region(camera.get(), offset_x, offset_y, width, height);
+
+        // Go through our settings and apply them to the camera
+        for (const auto& cfg : config["settings"]) {
+            auto key = cfg.first.as<std::string>();
+
+            // Skip the region keys as we handle them above
+            if (key == "Width" || key == "Height" || key == "OffsetX" || key == "OffsetY" || key == "PacketSize") {
```
Comment (CMurtagh-LGTM) [inline module/input/Camera/src/Camera.cpp:406]: I don't see `PacketSize` handled above?

File: module/input/Camera/src/Camera.cpp [hunk 4]
```diff
@@ -462,12 +228,408 @@ namespace module::input {
...
+            }
+            // The camera doesn't exist yet, so create a blank context and create a watchdog
+            else {
+                const auto camera_name = std::filesystem::path(config.path).stem().string();
+                cameras.emplace(std::make_pair(serial_number,
+                                               CameraContext{*this,
+                                                             config,
+                                                             camera_name,
+                                                             0,  // fourcc is set later
+                                                             XXH32(camera_name.data(), camera_name.size(), 0x4E55436C),
```
Comment (CMurtagh-LGTM) [inline module/input/Camera/src/Camera.cpp:593]: Magic number again

**Review by CMurtagh-LGTM — CHANGES_REQUESTED** [review]
I don't like the current cyclic dependency of CameraContext and Camera, I've marked those comments and will be making a follow up pr.

File: module/input/Camera/src/time_sync.cpp [hunk 5]
```diff
@@ -1,15 +1,27 @@
...
 
 namespace module::input {
 
     CameraContext::TimeCorrection sync_clocks(ArvDevice* device) {
-        using namespace std::chrono;
+        using namespace std::chrono;  // NOLINT(google-build-using-namespace) fine in function scope
+
+        // Check that camera supports ptp
+        auto* feature = arv_device_get_feature(device, "GevIEEE1588");
+        if (feature != nullptr) {
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply module/input/Camera/src/time_sync.cpp:15]: Not super important anyway as it's all USB3 vision which doesn't use PTP
That said PTP between the robots and using system clocks would be pretty cool

File: module/input/Camera/src/Camera.cpp (hunk 3 above)
Comment (TrentHouliston) [reply module/input/Camera/src/Camera.cpp:406]: True... that one is handled below.
It's for gige vision cameras

File: module/input/Camera/src/Camera.cpp (hunk 4 above)
Comment (TrentHouliston) [reply module/input/Camera/src/Camera.cpp:593]: Really we could just use 0, it is a magic number for the hashing algorithm.
It's just other places we've used this magic number fo xxhash and I suppose it was copied here

File: module/input/Camera/src/Camera.cpp (hunk 1 above)
Comment (TrentHouliston) [reply module/input/Camera/src/Camera.cpp:273]: Yeah this should only happen when a camera disappears/unplugged etc and needs to be rediscovered

File: module/input/Camera/src/time_sync.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (CMurtagh-LGTM) [reply module/input/Camera/src/time_sync.cpp:15]: So the feature will return `nullptr` then?

File: module/input/Camera/src/Camera.cpp (hunk 3 above)
Comment (CMurtagh-LGTM) [reply module/input/Camera/src/Camera.cpp:406]: Please update the comments to reflect this.

File: module/input/Camera/src/Camera.cpp (hunk 4 above)
Comment (CMurtagh-LGTM) [reply module/input/Camera/src/Camera.cpp:593]: Can it be a `constexpr` in a file imported from `shared/utility`?

File: module/input/Camera/src/time_sync.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply module/input/Camera/src/time_sync.cpp:15]: For a USB3 camera I believe so but I don't have one to test with

File: module/input/Camera/src/Camera.cpp [hunk 6]
```diff
@@ -416,7 +178,11 @@ namespace module::input {
                         auto Hwp_it = std::lower_bound(reactor.Hwps.begin(),
                                                        reactor.Hwps.end(),
                                                        std::make_pair(msg->timestamp, Eigen::Affine3d::Identity()),
-                                                       [](const auto& a, const auto& b) { return a.first < b.first; });
+                                                       [](const auto& a, const auto& b) {
+                                                           // There are no pointers. TODO: updates fix clang
+                                                           // NOLINTNEXTLINE(modernize-use-nullptr)
```
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply module/input/Camera/src/Camera.cpp:183]: Yes!

**Review by CMurtagh-LGTM — APPROVED** [review]

**Pull request merged**